src/server.rs

```diff
@@ -1,105 +1,285 @@
+#[cfg(unix)]
+use std::path::PathBuf;
+use std::{io::Result, net::SocketAddr, sync::Arc};
+
+#[cfg(all(unix, feature = "async-std"))]
+use async_std::os::unix::net::UnixListener;
+#[cfg(feature = "async-std")]
+use async_std::{
+    io::{Read, Write},
+    net::TcpListener,
+    prelude::*,
+    sync::Mutex,
+    task::spawn,
+};
+#[cfg(all(unix, feature = "tokio"))]
+use tokio::net::UnixListener;
+#[cfg(feature = "tokio")]
+use tokio::{
+    io::{AsyncRead as Read, AsyncWrite as Write},
+    net::TcpListener,
+    spawn,
+    stream::StreamExt,
+    sync::Mutex,
+};
+
+use crate::{
+    proto::{abci::*, decode, encode},
+    Consensus, Info, Mempool,
+};
+
+/// ABCI Server
+pub struct Server<C, M, I>
 where
-    A: Application + 'static + Send + Sync,
+    C: Consensus + 'static,
+    M: Mempool + 'static,
+    I: Info + 'static,
 {
-    env_logger::from_env(Env::default().default_filter_or("info"))
-        .try_init()
-        .ok();
-    let listener = TcpListener::bind(&addr).unwrap();
-    let incoming = listener.incoming();
-    let app = Arc::new(Mutex::new(app));
-    let server = incoming
-        .map_err(|err| panic!("Connection failed: {}", err))
-        .for_each(move |socket| {
-            info!("Got connection! {:?}", socket);
-            let framed = ABCICodec::new().framed(socket);
-            let (_writer, reader) = framed.split();
-            let app_instance = Arc::clone(&app);
-
-            let responses = reader.map(move |request| {
-                debug!("Got Request! {:?}", request);
-                respond(&app_instance, &request)
-            });
-
-            let writes = responses.fold(_writer, |writer, response| {
-                debug!("Return Response! {:?}", response);
-                writer.send(response)
-            });
-            tokio::spawn(writes.then(|_| Ok(())))
-        });
-
-    let mut rt = runtime::Builder::new()
-        .panic_handler(|_err| {
-            std::process::exit(1);
-            // std::panic::resume_unwind(err);
-        })
-        .build()
-        .unwrap();
-    rt.block_on(server).unwrap();
-    Ok(())
+    pub(crate) consensus: Arc<C>,
+    pub(crate) mempool: Arc<M>,
+    pub(crate) info: Arc<I>,
+    pub(crate) consensus_state: Arc<Mutex<ConsensusState>>,
 }
 
-fn respond<A>(app: &Arc<Mutex<A>>, request: &Request) -> Response
+impl<C, M, I> Server<C, M, I>
 where
-    A: Application + 'static + Send + Sync,
+    C: Consensus + 'static,
+    M: Mempool + 'static,
+    I: Info + 'static,
 {
-    let mut guard = app.lock().unwrap();
-    let app = guard.deref_mut();
+    /// Creates a new instance of [`Server`](struct.Server.html)
+    #[inline]
+    pub fn new(consensus: C, mempool: M, info: I) -> Self {
+        Self {
+            consensus: Arc::new(consensus),
+            mempool: Arc::new(mempool),
+            info: Arc::new(info),
+            consensus_state: Arc::new(Mutex::new(ConsensusState::default())),
```

Comment:
How about reflect the requirement of `Clone` in the trait, rather than construct an `Arc`, because the caller might already have `Arc`.

Reply:
`Clone` requirement does not mean that the caller have these objects inside `Arc`. It may be an expensive clone.